File: src/shipyard/core/ship_state.py
"""Durable in-flight ship state for resume-after-interruption.

When `shipyard ship` is in progress, the ship flow holds a lot of
context in memory: which PR is being shipped, which cloud runs were
dispatched, which targets have produced evidence, and the merge
policy that was in effect when the ship started. Today, all of that
lives only in Python memory. When the session dies (laptop closed,
OS restart, agent crash), that knowledge is lost and the next
session has to re-dispatch from scratch.

This module persists the minimum set of facts needed to resume an
in-flight ship:

    <state_dir>/ship/<pr>.json

    {
        "schema_version": 1,
        "pr": 224,
        "repo": "danielraffel/pulp",
        "branch": "feature/foo",
        "base_branch": "main",
        "head_sha": "abc1234...",
        "dispatched_runs": [
            {"target": "cloud", "provider": "namespace",
             "run_id": "24446948064", "status": "in_progress",
             "started_at": "...", "updated_at": "..."}
        ],
        "evidence_snapshot": {"macos": "pass", "linux": "pending"},
        "policy_signature": "0123abcd",
        "attempt": 2,
        "created_at": "...",
        "updated_at": "..."
    }

After merge/cancel/final-fail, the file is moved to an `archive/`
subdirectory, keyed by PR + timestamp, so historical context is
preserved for `shipyard evidence` inspection without cluttering the
active state directory.

This store is intentionally separate from:
- EvidenceStore (core/evidence.py) — merge-gating evidence keyed by
  branch+target. That store answers "is this PR mergeable?"; this
  store answers "what was I doing when my session died?".
- Queue (core/queue.py) — task queue keyed by job. The queue's
  stale-job recovery handles "process died," but it does not know
  which cloud run IDs correspond to which targets. This store
  persists exactly that linkage so a fresh session can pick up
  without re-dispatching.
"""

from __future__ import annotations

import contextlib
import hashlib
import json
import os
import tempfile
from dataclasses import dataclass, field, replace
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from pathlib import Path


SCHEMA_VERSION = 1


@dataclass(frozen=True)
class DispatchedRun:
    """A single remote run dispatched during a ship.

    For cloud targets, `run_id` is the GitHub Actions run ID. For
    SSH targets, `run_id` is Shipyard's internal job ID. `provider`
    distinguishes between cloud providers (e.g., "namespace",
    "github-hosted") or SSH transports ("ssh", "ssh-windows"). The
    `status` field is the last observed status from the poller:
    "pending", "in_progress", "completed", "failed", "cancelled".
    """

    target: str
    provider: str
    run_id: str
    status: str
    started_at: datetime
    updated_at: datetime
    attempt: int = 1

    def to_dict(self) -> dict[str, Any]:
        return {
            "target": self.target,
            "provider": self.provider,
            "run_id": self.run_id,
            "status": self.status,
            "started_at": self.started_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "attempt": self.attempt,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> DispatchedRun:
        return cls(
            target=d["target"],
            provider=d["provider"],
            run_id=str(d["run_id"]),
            status=d["status"],
            started_at=datetime.fromisoformat(d["started_at"]),
            updated_at=datetime.fromisoformat(d["updated_at"]),
            attempt=d.get("attempt", 1),
        )


@dataclass
class ShipState:
    """Durable record of an in-flight ship for a single PR.

    A ShipState is mutable in memory — the ship flow updates it as
    events happen (dispatch, poll update, evidence record, merge).
    Callers should save through ShipStateStore after every
    significant update so a crash loses at most one event.

    Human-context fields (`pr_url`, `pr_title`, `commit_subject`)
    are best-effort metadata so that `shipyard ship-state list` and
    `show` are self-describing — the operator doesn't have to
    remember what PR #42 was about to decide whether to resume it.
    """

    pr: int
    repo: str
    branch: str
    base_branch: str
    head_sha: str
    policy_signature: str
    pr_url: str = ""
    pr_title: str = ""
    commit_subject: str = ""
    dispatched_runs: list[DispatchedRun] = field(default_factory=list)
    evidence_snapshot: dict[str, str] = field(default_factory=dict)
    attempt: int = 1
    created_at: datetime = field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    updated_at: datetime = field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    schema_version: int = SCHEMA_VERSION

    def touch(self) -> None:
        """Bump `updated_at` to now. Call after every mutation."""
        self.updated_at = datetime.now(timezone.utc)

    def upsert_run(self, run: DispatchedRun) -> None:
        """Insert or replace a DispatchedRun matching on (target, run_id)."""
        for i, existing in enumerate(self.dispatched_runs):
            if existing.target == run.target and existing.run_id == run.run_id:
                self.dispatched_runs[i] = run
                self.touch()
                return
        self.dispatched_runs.append(run)
        self.touch()

    def get_run(self, target: str) -> DispatchedRun | None:
        """Return the most recent run for `target`, or None."""
        matches = [r for r in self.dispatched_runs if r.target == target]
        if not matches:
            return None
        return max(matches, key=lambda r: r.updated_at)

    def update_evidence(self, target: str, status: str) -> None:
        """Record a target's evidence status in the snapshot."""
        self.evidence_snapshot[target] = status
        self.touch()

    def is_sha_drift(self, current_sha: str) -> bool:
        """True if `current_sha` differs from the recorded head_sha."""
        return current_sha != self.head_sha

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "pr": self.pr,
            "repo": self.repo,
            "branch": self.branch,
            "base_branch": self.base_branch,
            "head_sha": self.head_sha,
            "policy_signature": self.policy_signature,
            "pr_url": self.pr_url,
            "pr_title": self.pr_title,
            "commit_subject": self.commit_subject,
            "dispatched_runs": [r.to_dict() for r in self.dispatched_runs],
            "evidence_snapshot": dict(self.evidence_snapshot),
            "attempt": self.attempt,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ShipState:
        return cls(
            pr=int(d["pr"]),
            repo=d["repo"],
            branch=d["branch"],
            base_branch=d["base_branch"],
            head_sha=d["head_sha"],
            policy_signature=d.get("policy_signature", ""),
            pr_url=d.get("pr_url", ""),
            pr_title=d.get("pr_title", ""),
            commit_subject=d.get("commit_subject", ""),
            dispatched_runs=[
                DispatchedRun.from_dict(r)
                for r in d.get("dispatched_runs", [])
            ],
            evidence_snapshot=dict(d.get("evidence_snapshot", {})),
            attempt=int(d.get("attempt", 1)),
            created_at=datetime.fromisoformat(d["created_at"]),
            updated_at=datetime.fromisoformat(d["updated_at"]),
            schema_version=int(d.get("schema_version", SCHEMA_VERSION)),
        )


@dataclass(frozen=True)
class PruneReport:
    """Summary of what `ShipStateStore.prune()` removed."""

    deleted_active: list[int] = field(default_factory=list)
    deleted_archived: list[str] = field(default_factory=list)

    @property
    def total(self) -> int:
        return len(self.deleted_active) + len(self.deleted_archived)

    def to_dict(self) -> dict[str, Any]:
        return {
            "deleted_active": list(self.deleted_active),
            "deleted_archived": list(self.deleted_archived),
            "total": self.total,
        }


@dataclass
class ShipStateStore:
    """Persists in-flight ship state, one file per PR.

    Active state lives at `<path>/<pr>.json`. On archive, the file
    is moved to `<path>/archive/<pr>-<timestamp>.json`. Writes are
    atomic (tempfile + os.replace) to prevent torn files on crash.
    Reads tolerate missing or corrupt files by returning None so the
    caller can treat a damaged state as "no state" and re-dispatch.
    """

    path: Path

    def __post_init__(self) -> None:
        self.path.mkdir(parents=True, exist_ok=True)
        self._archive_dir.mkdir(parents=True, exist_ok=True)

    @property
    def _archive_dir(self) -> Path:
        return self.path / "archive"

    def _state_path(self, pr: int) -> Path:
        return self.path / f"{pr}.json"

    def get(self, pr: int) -> ShipState | None:
        state_path = self._state_path(pr)
        if not state_path.exists():
            return None
        try:
            data = json.loads(state_path.read_text())
            return ShipState.from_dict(data)
        except (json.JSONDecodeError, KeyError, ValueError):
            # Corrupt file — treat as absent. A fresh ship will
            # overwrite it.
            return None

    def save(self, state: ShipState) -> None:
        """Write state atomically.

        The caller is responsible for updating `state.updated_at`
        (the mutation helpers `upsert_run`, `update_evidence`, and
        `touch` do this automatically). save() writes what it is
        given so tests and audit tools can back-date state files
        when needed.
        """
        state_path = self._state_path(state.pr)
        payload = json.dumps(state.to_dict(), indent=2) + "\n"
        # Atomic write: tempfile in the same directory (so rename is
        # atomic on the same filesystem), then os.replace to swing
        # the final name. Prevents half-written files on crash or
        # concurrent writers.
        fd, tmp_name = tempfile.mkstemp(
            prefix=f".{state.pr}.", suffix=".tmp", dir=str(self.path)
        )
        try:
            with os.fdopen(fd, "w") as f:
                f.write(payload)
            os.replace(tmp_name, state_path)
        except Exception:
            with contextlib.suppress(OSError):
                os.unlink(tmp_name)
            raise

    def delete(self, pr: int) -> None:
        """Remove the active state file for a PR, no archive."""
        state_path = self._state_path(pr)
        if state_path.exists():
            state_path.unlink()

    def archive(self, pr: int) -> Path | None:
        """Move active state to the archive dir. Returns the archived path.

        Returns None if no active state exists for this PR. Archive
        filenames include a UTC timestamp so repeated archives for
        the same PR do not collide (e.g., a PR shipped twice).
        """
        state_path = self._state_path(pr)
        if not state_path.exists():
            return None
        stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        dest = self._archive_dir / f"{pr}-{stamp}.json"
        os.replace(state_path, dest)
        return dest

    def list_active(self) -> list[ShipState]:
        """Return every active (non-archived) ShipState, sorted by PR."""
        states: list[ShipState] = []
        for entry in sorted(self.path.glob("*.json")):
            # pr filenames are pure integers; skip anything else
            # (e.g., stray temp files from an interrupted write).
            if not entry.stem.isdigit():
                continue
            state = self.get(int(entry.stem))
            if state is not None:
                states.append(state)
        return sorted(states, key=lambda s: s.pr)

    def list_archived(self) -> list[Path]:
        """Return every archived file, sorted by name (PR + timestamp)."""
        if not self._archive_dir.exists():
            return []
        return sorted(self._archive_dir.glob("*.json"))

    def prune(
        self,
        *,
        active_days: int = 14,
        archive_days: int = 30,
        closed_prs: set[int] | None = None,
        now: datetime | None = None,
    ) -> PruneReport:
        """Remove stale active and archived state files.

        Rules:
        - An **active** state file is deleted if its `updated_at` is
          older than `active_days` AND (closed_prs is provided and
          the PR is in it). Without a `closed_prs` set, active files
          are never auto-deleted — they may still be in flight.
        - An **archived** file is deleted if its filesystem mtime is
          older than `archive_days`. Archives are tombstones; once
          aged out they have no value.

        Returns a PruneReport listing what was deleted.
        """
        now = now or datetime.now(timezone.utc)
        active_cutoff = now - timedelta(days=active_days)
        archive_cutoff = now - timedelta(days=archive_days)

        deleted_active: list[int] = []
        deleted_archived: list[str] = []

        if closed_prs is not None:
            for state in self.list_active():
                if state.pr not in closed_prs:
                    continue
                if state.updated_at <= active_cutoff:
                    self.delete(state.pr)
                    deleted_active.append(state.pr)

        for archive_path in self.list_archived():
            mtime = datetime.fromtimestamp(
                archive_path.stat().st_mtime, tz=timezone.utc
            )
            if mtime <= archive_cutoff:
                archive_path.unlink()
                deleted_archived.append(archive_path.name)

        return PruneReport(
            deleted_active=deleted_active,
            deleted_archived=deleted_archived,
        )

    def archive_and_replace(
        self, state: ShipState, new_attempt: int | None = None
    ) -> ShipState:
        """Archive the existing state for this PR and start a new attempt.

        Used when a ship restarts for the same PR (e.g., user force-
        pushed and re-ran ship). Returns a fresh ShipState with the
        attempt counter incremented.
        """
        self.archive(state.pr)
        return replace(
            state,
            attempt=(new_attempt if new_attempt is not None else state.attempt + 1),
            dispatched_runs=[],
            evidence_snapshot={},
            created_at=datetime.now(timezone.utc),
            updated_at=datetime.now(timezone.utc),
        )


def compute_policy_signature(
    required_platforms: list[str], target_names: list[str], mode: str
) -> str:
    """Compute a stable digest of the merge policy inputs.

    If the user changes `.shipyard/config.toml` between ship start
    and resume — e.g. adds a new required platform — the signature
    no longer matches and the ship must refuse to resume. This
    prevents silently merging under a different policy than was in
    effect at dispatch time.
    """
    h = hashlib.sha256()
    h.update(b"platforms:")
    for p in sorted(required_platforms):
        h.update(p.encode("utf-8"))
        h.update(b"\x00")
    h.update(b"targets:")
    for t in sorted(target_names):
        h.update(t.encode("utf-8"))
        h.update(b"\x00")
    h.update(b"mode:")
    h.update(mode.encode("utf-8"))
    return h.hexdigest()[:16]

File: src/ship_state.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Deserializer, Serialize};
use sha2::{Digest, Sha256};

/// Schema version for durable ship-state files.
pub const SHIP_STATE_SCHEMA_VERSION: u32 = 1;

/// A single dispatched run tracked as part of an in-flight ship.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct DispatchedRun {
    /// Target name associated with the run.
    pub target: String,
    /// Provider or backend label.
    pub provider: String,
    /// Provider-specific run identifier.
    #[serde(deserialize_with = "deserialize_run_id")]
    pub run_id: String,
    /// Latest observed status.
    pub status: String,
    /// Timestamp when the run started.
    pub started_at: DateTime<Utc>,
    /// Timestamp when the run was last updated.
    pub updated_at: DateTime<Utc>,
    /// Attempt number for reruns.
    #[serde(default = "default_attempt")]
    pub attempt: u32,
    /// Optional last heartbeat timestamp for stale-run detection.
    #[serde(default)]
    pub last_heartbeat_at: Option<DateTime<Utc>>,
    /// Optional current phase name.
    #[serde(default)]
    pub phase: Option<String>,
    /// Whether this lane is merge-blocking.
    #[serde(default = "default_true")]
    pub required: bool,
}

impl DispatchedRun {
    /// Convert this run into the JSON shape emitted by the Python CLI.
    #[must_use]
    pub fn to_json_value(&self) -> serde_json::Value {
        serde_json::to_value(self).expect("DispatchedRun must serialize")
    }
}

/// Durable state for a single in-flight PR ship.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ShipState {
    /// Schema version for this state file.
    #[serde(default = "default_ship_state_schema_version")]
    pub schema_version: u32,
    /// Pull request number.
    pub pr: u64,
    /// Repository slug.
    pub repo: String,
    /// Head branch name.
    pub branch: String,
    /// Base branch name.
    pub base_branch: String,
    /// Recorded head SHA.
    pub head_sha: String,
    /// Merge-policy signature captured at dispatch time.
    #[serde(default)]
    pub policy_signature: String,
    /// Optional PR URL for self-describing state output.
    #[serde(default)]
    pub pr_url: String,
    /// Optional PR title for self-describing state output.
    #[serde(default)]
    pub pr_title: String,
    /// Optional commit subject for self-describing state output.
    #[serde(default)]
    pub commit_subject: String,
    /// Recorded remote runs.
    #[serde(default)]
    pub dispatched_runs: Vec<DispatchedRun>,
    /// Snapshot of evidence statuses by target.
    #[serde(default)]
    pub evidence_snapshot: BTreeMap<String, String>,
    /// Attempt number for this PR ship.
    #[serde(default = "default_attempt")]
    pub attempt: u32,
    /// Creation timestamp.
    pub created_at: DateTime<Utc>,
    /// Last update timestamp.
    pub updated_at: DateTime<Utc>,
}

impl ShipState {
    /// Construct a new in-flight ship state.
    #[must_use]
    pub fn new(
        pr: u64,
        repo: impl Into<String>,
        branch: impl Into<String>,
        base_branch: impl Into<String>,
        head_sha: impl Into<String>,
        policy_signature: impl Into<String>,
    ) -> Self {
        let now = Utc::now();
        Self {
            schema_version: SHIP_STATE_SCHEMA_VERSION,
            pr,
            repo: repo.into(),
            branch: branch.into(),
            base_branch: base_branch.into(),
            head_sha: head_sha.into(),
            policy_signature: policy_signature.into(),
            pr_url: String::new(),
            pr_title: String::new(),
            commit_subject: String::new(),
            dispatched_runs: Vec::new(),
            evidence_snapshot: BTreeMap::new(),
            attempt: default_attempt(),
            created_at: now,
            updated_at: now,
        }
    }

    /// Update the `updated_at` timestamp.
    pub fn touch(&mut self) {
        self.updated_at = Utc::now();
    }

    /// Insert or replace a run by `(target, run_id)`.
    pub fn upsert_run(&mut self, run: DispatchedRun) {
        if let Some(existing) = self
            .dispatched_runs
            .iter_mut()
            .find(|existing| existing.target == run.target && existing.run_id == run.run_id)
        {
            *existing = run;
        } else {
            self.dispatched_runs.push(run);
        }
        self.touch();
    }

    /// Return the most recently updated run for a target.
    #[must_use]
    pub fn get_run(&self, target: &str) -> Option<&DispatchedRun> {
        self.dispatched_runs
            .iter()
            .filter(|run| run.target == target)
            .max_by_key(|run| run.updated_at)
    }

    /// Return whether any run already exists for a target.
    #[must_use]
    pub fn has_target(&self, target: &str) -> bool {
        self.dispatched_runs.iter().any(|run| run.target == target)
    }

    /// Append a new run without deduplication.
    pub fn append_run(&mut self, run: DispatchedRun) {
        self.dispatched_runs.push(run);
        self.touch();
    }

    /// Update the saved evidence status for a target.
    pub fn update_evidence(&mut self, target: impl Into<String>, status: impl Into<String>) {
        self.evidence_snapshot.insert(target.into(), status.into());
        self.touch();
    }

    /// Return whether the recorded head SHA differs from the current SHA.
    #[must_use]
    pub fn is_sha_drift(&self, current_sha: &str) -> bool {
        current_sha != self.head_sha
    }
}

/// Report describing what a prune operation removed.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct PruneReport {
    /// Deleted active PR numbers.
    pub deleted_active: Vec<u64>,
    /// Deleted archived filenames.
    pub deleted_archived: Vec<String>,
}

impl PruneReport {
    /// Total number of deleted entries.
    #[must_use]
    pub fn total(&self) -> usize {
        self.deleted_active.len() + self.deleted_archived.len()
    }
}

/// Persistent store for active and archived ship-state files.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ShipStateStore {
    path: PathBuf,
}

impl ShipStateStore {
    /// Open a state store at the given path.
    pub fn new(path: PathBuf) -> Result<Self, std::io::Error> {
        fs::create_dir_all(&path)?;
        fs::create_dir_all(path.join("archive"))?;
        Ok(Self { path })
    }

    /// Backing path of the store.
    #[must_use]
    pub fn path(&self) -> &Path {
        &self.path
    }

    /// Return the active state path for a PR.
    #[must_use]
    pub fn state_path(&self, pr: u64) -> PathBuf {
        self.path.join(format!("{pr}.json"))
    }

    /// Return the archive directory path.
    #[must_use]
    pub fn archive_dir(&self) -> PathBuf {
        self.path.join("archive")
    }

    /// Load an active state for a PR.
    #[must_use]
    pub fn get(&self, pr: u64) -> Option<ShipState> {
        let path = self.state_path(pr);
        let contents = fs::read_to_string(path).ok()?;
        serde_json::from_str(&contents).ok()
    }

    /// Save a state atomically.
    pub fn save(&self, state: &ShipState) -> Result<(), Box<dyn std::error::Error>> {
        let payload = serde_json::to_string_pretty(state)?;
        let temp = tempfile::NamedTempFile::new_in(&self.path)?;
        fs::write(temp.path(), format!("{payload}\n"))?;
        temp.persist(self.state_path(state.pr))?;
        Ok(())
    }

    /// Delete an active state file.
    pub fn delete(&self, pr: u64) -> Result<(), std::io::Error> {
        let path = self.state_path(pr);
        if path.exists() {
            fs::remove_file(path)?;
        }
        Ok(())
    }

    /// Move an active state into the archive directory.
    pub fn archive(&self, pr: u64) -> Result<Option<PathBuf>, std::io::Error> {
        let source = self.state_path(pr);
        if !source.exists() {
            return Ok(None);
        }
        let stamp = Utc::now().format("%Y%m%dT%H%M%SZ");
        let dest = self.archive_dir().join(format!("{pr}-{stamp}.json"));
        fs::rename(source, &dest)?;
        Ok(Some(dest))
    }

    /// Return active states sorted by PR number.
    pub fn list_active(&self) -> Vec<ShipState> {
        let mut states = Vec::new();
        if let Ok(entries) = fs::read_dir(&self.path) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.parent() == Some(self.archive_dir().as_path()) {
                    continue;
                }
                if path.extension().and_then(std::ffi::OsStr::to_str) != Some("json") {
                    continue;
                }
                if !path
                    .file_stem()
                    .and_then(std::ffi::OsStr::to_str)
                    .is_some_and(|stem| stem.chars().all(|ch| ch.is_ascii_digit()))
                {
                    continue;
                }
                if let Ok(contents) = fs::read_to_string(&path)
                    && let Ok(state) = serde_json::from_str::<ShipState>(&contents)
                {
                    states.push(state);
                }
            }
        }
        states.sort_by_key(|state| state.pr);
        states
    }

    /// Return archived state file paths sorted by filename.
    pub fn list_archived(&self) -> Vec<PathBuf> {
        let mut archived = Vec::new();
        if let Ok(entries) = fs::read_dir(self.archive_dir()) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().and_then(std::ffi::OsStr::to_str) == Some("json") {
                    archived.push(path);
                }
            }
        }
        archived.sort();
        archived
    }

    /// Archive the current state for a PR and create a fresh attempt record.
    pub fn archive_and_replace(
        &self,
        state: &ShipState,
        new_attempt: Option<u32>,
    ) -> Result<ShipState, Box<dyn std::error::Error>> {
        let _ = self.archive(state.pr)?;
        let now = Utc::now();
        Ok(ShipState {
            attempt: new_attempt.unwrap_or(state.attempt + 1),
            dispatched_runs: Vec::new(),
            evidence_snapshot: BTreeMap::new(),
            created_at: now,
            updated_at: now,
            ..state.clone()
        })
    }
}

/// Compute a stable digest of merge-policy inputs.
#[must_use]
pub fn compute_policy_signature(
    required_platforms: &[String],
    target_names: &[String],
    mode: &str,
) -> String {
    let mut hasher = Sha256::new();
    hasher.update(b"platforms:");
    let mut platforms = required_platforms.to_vec();
    platforms.sort();
    for platform in platforms {
        hasher.update(platform.as_bytes());
        hasher.update([0]);
    }
    hasher.update(b"targets:");
    let mut targets = target_names.to_vec();
    targets.sort();
    for target in targets {
        hasher.update(target.as_bytes());
        hasher.update([0]);
    }
    hasher.update(b"mode:");
    hasher.update(mode.as_bytes());
    let digest = hasher.finalize();
    hex::encode(&digest[..8])
}

fn default_attempt() -> u32 {
    1
}

fn default_true() -> bool {
    true
}

fn default_ship_state_schema_version() -> u32 {
    SHIP_STATE_SCHEMA_VERSION
}

fn deserialize_run_id<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    let value = serde_json::Value::deserialize(deserializer)?;
    match value {
        serde_json::Value::String(text) => Ok(text),
        serde_json::Value::Number(number) => Ok(number.to_string()),
        other => Err(serde::de::Error::custom(format!(
            "run_id must be a string or number, got {other}"
        ))),
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use chrono::{Duration, TimeZone, Utc};

    use super::{DispatchedRun, ShipState, ShipStateStore, compute_policy_signature};

    fn sample_state(pr: u64, sha: &str) -> ShipState {
        ShipState::new(
            pr,
            "danielraffel/pulp",
            "feature/test",
            "main",
            sha,
            "policy0001",
        )
    }

    fn sample_run(target: &str, run_id: &str) -> DispatchedRun {
        let now = Utc::now();
        DispatchedRun {
            target: target.to_owned(),
            provider: "namespace".to_owned(),
            run_id: run_id.to_owned(),
            status: "in_progress".to_owned(),
            started_at: now,
            updated_at: now,
            attempt: 1,
            last_heartbeat_at: None,
            phase: None,
            required: true,
        }
    }

    #[test]
    fn dispatched_run_roundtrip_accepts_numeric_run_id() {
        let value = serde_json::json!({
            "target": "cloud",
            "provider": "namespace",
            "run_id": 24_446_948_064_u64,
            "status": "in_progress",
            "started_at": "2026-04-15T10:00:00+00:00",
            "updated_at": "2026-04-15T10:00:00+00:00"
        });
        let run: DispatchedRun = serde_json::from_value(value).expect("run should deserialize");
        assert_eq!(run.run_id, "24446948064");
        assert_eq!(run.attempt, 1);
        assert!(run.required);
    }

    #[test]
    fn ship_state_roundtrip_preserves_optional_fields() {
        let mut state = sample_state(224, "abc1234");
        state.pr_url = "https://github.com/danielraffel/pulp/pull/224".to_owned();
        state.pr_title = "Fix ARA controller".to_owned();
        state.commit_subject = "ara: out-of-line destructor".to_owned();
        state.upsert_run(sample_run("cloud", "99999"));
        state.update_evidence("macos", "pass");

        let restored: ShipState =
            serde_json::from_value(serde_json::to_value(&state).expect("serialize"))
                .expect("deserialize");
        assert_eq!(restored, state);
    }

    #[test]
    fn get_run_returns_most_recent_match() {
        let mut state = sample_state(1, "abc");
        let older = Utc::now() - Duration::minutes(10);
        let newer = Utc::now();
        state.dispatched_runs.push(DispatchedRun {
            updated_at: older,
            started_at: older,
            ..sample_run("cloud", "111")
        });
        state.dispatched_runs.push(DispatchedRun {
            updated_at: newer,
            started_at: newer,
            ..sample_run("cloud", "222")
        });
        assert_eq!(
            state.get_run("cloud").map(|run| run.run_id.as_str()),
            Some("222")
        );
    }

    #[test]
    fn store_save_get_list_and_archive_roundtrip() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = ShipStateStore::new(temp.path().join("ship")).expect("store");
        let mut state = sample_state(42, "abc1234");
        state.upsert_run(sample_run("cloud", "999"));
        store.save(&state).expect("save");

        let restored = store.get(42).expect("state should exist");
        assert_eq!(restored.pr, 42);
        assert_eq!(
            store
                .list_active()
                .iter()
                .map(|item| item.pr)
                .collect::<Vec<_>>(),
            vec![42]
        );

        let archived = store
            .archive(42)
            .expect("archive call")
            .expect("archive path");
        assert!(archived.exists());
        assert!(store.get(42).is_none());
        assert_eq!(store.list_archived().len(), 1);
    }

    #[test]
    fn list_active_ignores_corrupt_and_non_integer_files() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = ShipStateStore::new(temp.path().join("ship")).expect("store");
        store.save(&sample_state(7, "abc")).expect("save");
        fs::write(store.path().join("notapr.json"), "{}").expect("write stray");
        fs::write(store.state_path(21), "{broken").expect("write corrupt");
        let prs = store
            .list_active()
            .iter()
            .map(|state| state.pr)
            .collect::<Vec<_>>();
        assert_eq!(prs, vec![7]);
    }

    #[test]
    fn archive_and_replace_increments_attempt_and_clears_live_fields() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = ShipStateStore::new(temp.path().join("ship")).expect("store");
        let mut state = sample_state(30, "abc");
        state.attempt = 1;
        state.upsert_run(sample_run("cloud", "999"));
        state.update_evidence("macos", "pass");
        store.save(&state).expect("save");

        let fresh = store
            .archive_and_replace(&state, None)
            .expect("archive and replace");
        assert_eq!(fresh.attempt, 2);
        assert!(fresh.dispatched_runs.is_empty());
        assert!(fresh.evidence_snapshot.is_empty());
        assert_eq!(store.list_archived().len(), 1);
    }

    #[test]
    fn compute_policy_signature_is_stable_and_changes_with_inputs() {
        let a = compute_policy_signature(
            &["macos".to_owned(), "linux".to_owned(), "windows".to_owned()],
            &["mac".to_owned(), "ubuntu".to_owned()],
            "default",
        );
        let b = compute_policy_signature(
            &["windows".to_owned(), "macos".to_owned(), "linux".to_owned()],
            &["ubuntu".to_owned(), "mac".to_owned()],
            "default",
        );
        let c = compute_policy_signature(&["macos".to_owned()], &["mac".to_owned()], "strict");
        assert_eq!(a, b);
        assert_ne!(a, c);
        assert_eq!(a.len(), 16);
    }

    #[test]
    fn legacy_state_without_human_context_fields_still_loads() {
        let value = serde_json::json!({
            "schema_version": 1,
            "pr": 224,
            "repo": "danielraffel/pulp",
            "branch": "feature/test",
            "base_branch": "main",
            "head_sha": "abc1234",
            "policy_signature": "policy0001",
            "dispatched_runs": [],
            "evidence_snapshot": {},
            "attempt": 1,
            "created_at": "2026-04-15T10:00:00+00:00",
            "updated_at": "2026-04-15T10:00:00+00:00"
        });
        let state: ShipState = serde_json::from_value(value).expect("legacy state");
        assert_eq!(state.pr_url, "");
        assert_eq!(state.pr_title, "");
        assert_eq!(state.commit_subject, "");
    }

    #[test]
    fn save_is_atomic_and_leaves_no_named_tempfiles() {
        let temp = tempfile::tempdir().expect("tempdir");
        let store = ShipStateStore::new(temp.path().join("ship")).expect("store");
        store.save(&sample_state(55, "abc")).expect("save");
        let strays = fs::read_dir(store.path())
            .expect("read dir")
            .filter_map(Result::ok)
            .map(|entry| entry.file_name().to_string_lossy().into_owned())
            .filter(|name| name.starts_with('.'))
            .collect::<Vec<_>>();
        assert!(strays.is_empty(), "unexpected temp files: {strays:?}");
    }

    #[test]
    fn touch_and_sha_drift_behave_as_expected() {
        let mut state = sample_state(1, "abc");
        let original = Utc
            .with_ymd_and_hms(2026, 1, 1, 0, 0, 0)
            .single()
            .expect("valid date");
        state.updated_at = original;
        state.touch();
        assert!(state.updated_at >= original);
        assert!(!state.is_sha_drift("abc"));
        assert!(state.is_sha_drift("def"));
    }
}
